homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,117 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Iterable
+from dataclasses import dataclass
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.climate.intent import INTENT_GET_TEMPERATURE
+from homeassistant.components.weather.intent import INTENT_GET_WEATHER
+from homeassistant.core import Context, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.util.json import JsonObjectType
+
+from . import intent
+
+_LOGGER = logging.getLogger(__name__)
+
+IGNORE_INTENTS = [
+    intent.INTENT_NEVERMIND,
+    intent.INTENT_GET_STATE,
+    INTENT_GET_WEATHER,
+    INTENT_GET_TEMPERATURE,
+]
+
+
+@dataclass(slots=True)
+class ToolInput:
+    """Tool input to be processed."""
+
+    tool_name: str
+    tool_args: dict[str, Any]
+    platform: str
+    context: Context | None
+    user_prompt: str | None
+    language: str | None
+    assistant: str | None
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str | None = None
+    parameters: vol.Schema = vol.Schema({})
+
+    @abstractmethod
+    async def async_call(
+        self, hass: HomeAssistant, tool_input: ToolInput
+    ) -> JsonObjectType:
+        """Call the tool."""
+        raise NotImplementedError
+
+    def __repr__(self) -> str:
+        """Represent a string of a Tool."""
+        return f"<{self.__class__.__name__} - {self.name}>"
+
+
+@callback
+def async_get_tools(hass: HomeAssistant) -> Iterable[Tool]:
+    """Return a list of LLM tools."""
+    for intent_handler in intent.async_get(hass):
+        if intent_handler.intent_type not in IGNORE_INTENTS:
+            yield IntentTool(intent_handler)
+
+
+@callback
+async def async_call_tool(hass: HomeAssistant, tool_input: ToolInput) -> JsonObjectType:
```

Comment:
I guess this is the only thing that still feels a bit weird…

We are mutating a passed in object.

Reply:
We can fix that later

Reply:
Should I create a copy of the object, or change the approach?

Reply:
Fixed